### PR TITLE
Authentication specs #1

### PR DESCRIPTION
**Pull request by tomglvng**
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/DTO/AppAuthorizationSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/DTO/AsymmetricKeysSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Model/AuthenticationScopeSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/ValueObject/ScopeListSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/AsymmetricKeysGeneratorSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateAccessTokenSpec.php
- src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateJsonWebTokenSpec.php

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class OpenIdKeysNotFoundExceptionSpec  extends ObjectBehavior
+{
+    public function let(): void
+    {
+        $this->beConstructedWith('an_app_id', 1234, ['a_scope', 'another_scope']);
+    }
+
+    public function it_is_initializable(): void
+    {
+        $this->shouldHaveType(OpenIdKeysNotFoundException::class);
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php:22]: and :
```PHP
       $this->shouldHaveType(\Exception::class);
```

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class OpenIdKeysNotFoundExceptionSpec  extends ObjectBehavior
+{
+    public function let(): void
+    {
+        $this->beConstructedWith('an_app_id', 1234, ['a_scope', 'another_scope']);
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php:17]: Shouldn't it be empty ?

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+class UserConsentRequiredExceptionSpec extends ObjectBehavior
+{
+    public function let(): void
+    {
+        $this->beConstructedWith('an_app_id', 1234, ['a_scope', 'another_scope']);
+    }
+
+    public function it_is_initializable(): void
+    {
+        $this->shouldHaveType(UserConsentRequiredException::class);
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php:23]:
```PHP
$this->shouldHaveType(\Exception::class);
```

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class UserConsentRequiredExceptionSpec extends ObjectBehavior
+{
+    public function let(): void
+    {
+        $this->beConstructedWith('an_app_id', 1234, ['a_scope', 'another_scope']);
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php:18]: Side note : parent constructor not called

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateAccessTokenSpec.php [hunk 5]
```diff
@@ -247,4 +251,96 @@ public function it_processes_only_valid_auth_code(
...
+        $getUserConsentedAuthenticationScopesQuery->execute(2, 'client_id_1234')
+            ->willReturn([]);
+
+        $authCode->getScope()
+            ->willReturn('delete_products');
+        $authCode->getData()
+            ->willReturn($pimUser);
+
+        $getAccessTokenQuery->execute('client_id_1234', ['scope1', 'scope2'])->willReturn(null);
+
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateAccessTokenSpec.php:291]:
```suggestion
```

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php (hunk 2 above)
Comment (tomglvng) [reply src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/OpenIdKeysNotFoundExceptionSpec.php:17]: Yes ;)

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php (hunk 4 above)
Comment (tomglvng) [reply src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Domain/Apps/Exception/UserConsentRequiredExceptionSpec.php:18]: I added it

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateJsonWebTokenSpec.php [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    public function it_creates_jwt_token(
+        GetAsymmetricKeysQueryInterface $getAsymmetricKeysQuery,
+        Clock $clock,
+        PimUrl $pimUrl
+    ): void {
+        $getAsymmetricKeysQuery->execute()->willReturn(AsymmetricKeys::create($this->publicKey, $this->privateKey));
+        $clock->now()->willReturn($this->now);
+        $pimUrl->getPimUrl()->willReturn($this->pimUrl);
```
Comment (RiffVentura) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Infrastructure/Apps/OAuth/CreateJsonWebTokenSpec.php:61]: all 3 test repeats it : could be moved into let method

**Review by RiffVentura — APPROVED** [review]

**Review by pchasle — APPROVED** [review]
:fireworks:

**Pull request merged**